### PR TITLE
Updates for new auth_state, added logout token revocation

### PR DESCRIPTION
**Pull request by NickolausDS**
The new changes make the Globus OAuthenticator compatible with the newest version of Jupyterhub. This includes passing user data through the proper channels to be encrypted in the database. 

In addition, there are a couple added logout features. Admins can now set a URL to logout of their identity provider, as well as an added security measure to set client tokens to automatically be revoked on logout.

### CODE REVIEW

File: oauthenticator/globus.py [hunk 1]
```diff
@@ -31,36 +32,6 @@ class GlobusMixin(OAuth2Mixin):
...
-            self.set_login_cookie(user)
-            self.redirect(url_path_join(self.hub.server.base_url, 'home'))
-        else:
-            raise web.HTTPError(403)
-
-    def set_globus_data(self, user):
-        user.auth_state = {
-            'globus_data': self.authenticator.globus_data
-        }
-        self.db.commit()
```
Comment (NickolausDS) [inline oauthenticator/globus.py:63]: No longer needed, now that passing data to `auth_state` is supported.

File: oauthenticator/globus.py [hunk 2]
```diff
@@ -72,22 +43,58 @@ class GlobusLoginHandler(OAuthLoginHandler, GlobusMixin):
...
         redirect_uri = self.authenticator.get_callback_url(self)
         client = self.authenticator.globus_portal_client()
+        state = self.get_state()
+        self.set_state_cookie(state)
         client.oauth2_start_flow(
             redirect_uri,
             requested_scopes=' '.join(self.authenticator.scope),
-            refresh_tokens=self.authenticator.allow_refresh_tokens
+            refresh_tokens=self.authenticator.allow_refresh_tokens,
+            state=state
```
Comment (NickolausDS) [inline oauthenticator/globus.py:52]: `state` wasn't originally checked for in the older oauthenticator. Now that it's built in, we can pass it through in the first hop and it will automatically be checked before getting to the `authenticate` method.

File: oauthenticator/globus.py [hunk 3]
```diff
@@ -128,36 +135,41 @@ def _allow_refresh_tokens_default(self):
...
+        return os.getenv('LOGOUT_REDIRECT_URL', '')
+
+    revoke_tokens_on_logout = Bool(
+        help="""Revoke tokens so they cannot be used again. Single-user servers
+        MUST be restarted after logout in order to get a fresh working set of
+        tokens."""
+    ).tag(config=True)
+
+    def _revoke_tokens_on_logout_default(self):
+        return False
```
Comment (NickolausDS) [inline oauthenticator/globus.py:151]: New features for logout!

`revoke_token_on_logout` is set to false by default due to a rather unintuitive user-experience of needing to manually shut down their server.

File: oauthenticator/globus.py [hunk 4]
```diff
@@ -195,7 +200,29 @@ def authenticate(self, handler, data=None):
...
+        return {
+            'name': username,
+            'auth_state': {
+                'client_id': self.client_id,
+                'tokens': {
+                    tok: v for tok, v in tokens.by_resource_server.items()
+                    if tok not in self.exclude_tokens
+                },
+            }
+        }
```
Comment (briedel) [inline oauthenticator/globus.py]: Did they change the return type, such that this passes the token to their sqlite DB?

File: oauthenticator/globus.py (hunk 1 above)
Comment (briedel) [reply oauthenticator/globus.py:63]: What is the `auth_state` here? The return of the `authenticate` method?

File: oauthenticator/globus.py (hunk 2 above)
Comment (briedel) [reply oauthenticator/globus.py:52]: What is `state`?

File: oauthenticator/tests/test_globus.py [hunk 5]
```diff
@@ -54,14 +54,27 @@ def globus_client(client):
...
+        @gen.coroutine
+        def get_auth_state(self):
+            return {'tokens': {}}
+
+        save_auth_state = Mock()
+
+        name = 'wash'
+    return User()
+
+
```
Comment (briedel) [inline oauthenticator/tests/test_globus.py]: What is the used for?

File: oauthenticator/tests/test_globus.py [hunk 6]
```diff
@@ -54,14 +54,27 @@ def globus_client(client):
...
     authenticator = GlobusOAuthenticator()
     handler = globus_client.handler_for_user(user_model('wash'))
-    name = yield authenticator.authenticate(handler)
-    assert name == 'wash'
-    assert list(authenticator.globus_data['tokens'].keys()) == \
-        ['transfer.api.globus.org']
+    data = yield authenticator.authenticate(handler)
+    assert data['name'] == 'wash'
+    tokens = list(data['auth_state']['tokens'].keys())
+    assert tokens == ['transfer.api.globus.org']
```
Comment (briedel) [inline oauthenticator/tests/test_globus.py:64]: Are these changes needed because of the change in the return type of `authenticate`?

File: oauthenticator/globus.py (hunk 1 above)
Comment (NickolausDS) [reply oauthenticator/globus.py:63]: `auth_state` is brand new and mainly the reason for the updates. The Jupyterhub folks didn't like sensitive material being non-encrypted within the Jupyterhub database at rest. They changed `auth_state` from a regular JSON field to an encrypted JSOS field, and data is saved by returning a dict here from the `authenticate` method instead of calling out to the `user` class specifically. 

Oauthenticator just had another update this morning, so this may change: https://github.com/jupyterhub/oauthenticator/issues/110

**EDIT**: All changes here reflect now build on the recent changes upstream.

File: oauthenticator/globus.py (hunk 2 above)
Comment (NickolausDS) [reply oauthenticator/globus.py:52]: From https://docs.globus.org/api/auth/developer-guide/, the `state` variable is: 

> An opaque value that is returned to the client in the authorization response. The client should use this value to prevent cross-site request forgery as described in https://tools.ietf.org/html/rfc6749#section-10.12.

It really doesn't change the end result in any way, but it makes the login process more secure.

File: oauthenticator/globus.py (hunk 4 above)
Comment (NickolausDS) [reply oauthenticator/globus.py]: They did. Where it used to just be `username`, it's now a dict. Not long ago, I think only one or two authenticators were returning this dict. As of this morning, it looks like all of them are. I'll have to revisit their changes and amend the ones here.

File: oauthenticator/tests/test_globus.py (hunk 5 above)
Comment (NickolausDS) [reply oauthenticator/tests/test_globus.py]: The `User` fixture is a Jupyterhub user mock, intended to track when `get_auth_state` and `save_auth_state` were called. It's only used for the one test on token revocation on logout.

File: oauthenticator/tests/test_globus.py (hunk 6 above)
Comment (NickolausDS) [reply oauthenticator/tests/test_globus.py:64]: Yes.

File: oauthenticator/tests/test_globus.py [hunk 7]
```diff
@@ -107,9 +103,68 @@ def test_token_exclusion(globus_client, mock_globus_sdk):
...
+@mark.gen_test
+def test_logout_revokes_tokens(monkeypatch):
+
+    class User:
+        @gen.coroutine
+        def get_auth_state(self):
+            return {'tokens': {}}
+
+        save_auth_state = Mock()
+        name = 'Wash'
```
Comment (NickolausDS) [inline oauthenticator/tests/test_globus.py:153]: I went ahead and moved the `User` class here to make it more obvious. For now, this is the only test that should ever need it, due to it being the only one that calls out directly to its methods.

**Comment by NickolausDS** [comment]
Sending pull request to upstream...
